### PR TITLE
API-2243 - Rate limit with Mongo

### PR DESCRIPTION
**Pull request by tom59**

### CODE REVIEW

File: project/HmrcBuild.scala [hunk 1]
```diff
@@ -46,7 +47,8 @@ object HmrcBuild extends Build {
     "org.mockito" % "mockito-core" % "2.3.0" % "test,it",
     "com.github.tomakehurst" % "wiremock" % "2.1.12" % "test,it",
     "org.scalatestplus.play" %% "scalatestplus-play" % "2.0.0-M1" %  "test,it",
-    "org.scalaj" %% "scalaj-http" % "2.3.0" % "test,it"
+    "org.scalaj" %% "scalaj-http" % "2.3.0" % "test,it",
+     "uk.gov.hmrc" %% "reactivemongo-test" % "2.0.0"
```
Comment (gokyo) [inline project/HmrcBuild.scala]: empty space

File: app/uk/gov/hmrc/apigateway/util/Time.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.apigateway.util
+
+import org.joda.time.DateTime
+
+object Time {
+  def minutesSinceEpoch(): Int = (DateTime.now().getMillis / 1000 / 60).toInt
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/util/Time.scala]: I would use 
```
TimeUnit.MILLISECONDS.toMinutes(System.currentTimeMillis())
```

File: app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala [hunk 3]
```diff
@@ -59,16 +59,17 @@ class UserRestrictedEndpointFilter @Inject()
   private def validateRequestAndSwapToken(requestHeader: RequestHeader, proxyRequest: ProxyRequest): Future[RequestHeader] = {
     getAuthority(proxyRequest) flatMap { authority =>
 
-      val validateSubscriptions: Future[Unit] = for {
+      val validateSubscriptionsAndRateLimit: Future[Unit] = for {
         application <- getApplicationByClientId(authority.delegatedAuthority.clientId)
         _ <- applicationService.validateApplicationIsSubscribedToApi(application.id.toString,
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: what if we run these two steps in parallel (UserRestrictedEndpointFilter.scala)?
- validate application subscriptions
- validate application rate limit

File: app/uk/gov/hmrc/apigateway/play/filter/ApplicationRestrictedEndpointFilter.scala [hunk 4]
```diff
@@ -68,6 +68,7 @@ class ApplicationRestrictedEndpointFilter @Inject()
       app <- applicationFuture
       _ <- applicationService.validateApplicationIsSubscribedToApi(app.id.toString,
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/play/filter/ApplicationRestrictedEndpointFilter.scala]: what if we run these two steps in parallel (ApplicationRestrictedEndpointFilter.scala)?
- validate application subscriptions
- validate application rate limit

File: app/uk/gov/hmrc/apigateway/exception/GatewayError.scala [hunk 5]
```diff
@@ -42,6 +42,8 @@ object GatewayError {
 
   case class InvalidSubscription() extends GatewayError("RESOURCE_FORBIDDEN", "The application is not subscribed to the API which it is attempting to invoke")
 
+  case class ThrottledOut() extends GatewayError("MESSAGE_THROTTLED_OUT", "The request for the API is throttled as you have exceeded your quota.")
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/exception/GatewayError.scala]: shall we specify the threshold value in the error message? Or do we assume the user should know it?

Comment (tom59) [inline app/uk/gov/hmrc/apigateway/exception/GatewayError.scala]: the rate is approximate and per datacenter so I think it is better not to publish the rate in the message.
Also, we can not do it until wso2 is gone as it is the same message than wso2

File: app/uk/gov/hmrc/apigateway/repository/RateLimitRepository.scala [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future.sequence
+
+case class RateLimitCounter(clientId: String, minutesSinceEpoch: Int, createdAt: LocalDateTime = now(), count: Int = 1)
+
+@Singleton
+class RateLimitRepository @Inject()(val reactiveMongoApi: ReactiveMongoApi) {
+
+  implicit val dateTimeFormat = uk.gov.hmrc.mongo.json.ReactiveMongoFormats.localDateTimeFormats
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/repository/RateLimitRepository.scala]: What about this change?

```
import `uk.gov.hmrc.mongo.json.ReactiveMongoFormats`
...
implicit val dateTimeFormat = ReactiveMongoFormats.localDateTimeFormats
```

File: test/uk/gov/hmrc/apigateway/repository/RateLimitRepositorySpec.scala [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      }
+    }
+
+    "reset the threshold when a minute has passed" in {
+
+      await(underTest.validateAndIncrement("clientId", 2))
+      await(underTest.validateAndIncrement("clientId", 2))
+      setCurrentMillisFixed(now().plusMinutes(1).getMillis)
+
+      await(underTest.validateAndIncrement("clientId", 2))
```
Comment (gokyo) [inline test/uk/gov/hmrc/apigateway/repository/RateLimitRepositorySpec.scala]:
```
val result = await(underTest.validateAndIncrement("clientId", 2))

result shouldBe ()
```

**Pull request merged**